### PR TITLE
Little Refactors

### PR DESCRIPTION
**Pull request by Kixiron**
* Made code more consistent in formatting
* More closely held to Rust 2018 idioms 
* Readability improvements

### CODE REVIEW

File: src/docbuilder/mod.rs [hunk 1]
```diff
@@ -56,40 +52,43 @@ impl DocBuilder {
...
-                               WHERE crates.id = releases.crate_id",
-                   &[])
-            .unwrap() {
+        for row in &conn
+            .query(
+                "SELECT name, version FROM crates, releases \
+                 WHERE crates.id = releases.crate_id",
+                &[],
+            )
+            .unwrap()
```
Comment (jyn514) [inline src/docbuilder/mod.rs:67]: Can you return this error instead of unwrapping?

File: src/docbuilder/rustwide_builder.rs [hunk 2]
```diff
@@ -108,7 +108,7 @@ impl RustwideBuilder {
 
         let mut targets_to_install = TARGETS
             .iter()
-            .map(|t| t.to_string())
+            .map(|t| (*t).to_string())
```
Comment (jyn514) [inline src/docbuilder/rustwide_builder.rs:111]: This dereference is redundant, probably inserted by clippy.
```suggestion
            .map(|t| t.to_string())
```

File: src/docbuilder/rustwide_builder.rs [hunk 3]
```diff
@@ -215,7 +215,7 @@ impl RustwideBuilder {
                     let file_name = if versioned {
                         format!("{}-{}.{}", segments[1], rustc_version, segments[0])
                     } else {
-                        file.to_string()
+                        (*file).to_string()
```
Comment (jyn514) [inline src/docbuilder/rustwide_builder.rs:218]: Redundant dereference
```suggestion
                        file.to_string()
```

File: src/utils/daemon.rs [hunk 4]
```diff
@@ -25,13 +25,16 @@ use ::{
 };
 
 pub fn start_daemon(background: bool) {
+    const CRATE_VARIABLES: [&str; 4] = [
+        "CRATESFYI_PREFIX",
+        "CRATESFYI_PREFIX",
+        "CRATESFYI_GITHUB_USERNAME",
+        "CRATESFYI_GITHUB_ACCESSTOKEN"
+    ];
```
Comment (jyn514) [inline src/utils/daemon.rs:33]:
```suggestion
    const CRATE_VARIABLES: [&str; 3] = [
        "CRATESFYI_PREFIX",
        "CRATESFYI_GITHUB_USERNAME",
        "CRATESFYI_GITHUB_ACCESSTOKEN"
    ];
```

File: src/utils/github_updater.rs [hunk 5]
```diff
@@ -109,27 +108,29 @@ fn get_github_fields(path: &str) -> Result<GitHubFields> {
...
-                             reponame
-                         }))
+
+            Some(format!(
+                "{}/{}",
+                username,
+                if reponame.ends_with(".git") {
+                   reponame.split(".git").next().unwrap()
+                } else {
+                    reponame
```
Comment (jyn514) [inline src/utils/github_updater.rs:129]: Can you turn this into a temporary variable so it doesn't make this expression really big?

File: src/utils/release_activity_updater.rs [hunk 6]
```diff
@@ -1,42 +1,50 @@
...
 
 pub fn update_release_activity() -> Result<()> {
-
     let conn = connect_db()?;
-    let mut dates = Vec::new();
-    let mut crate_counts = Vec::new();
-    let mut failure_counts = Vec::new();
+    let mut dates = Vec::with_capacity(30);
+    let mut crate_counts = Vec::with_capacity(30);
+    let mut failure_counts = Vec::with_capacity(30);
```
Comment (jyn514) [inline src/utils/release_activity_updater.rs:11]: Ooh this is nice, I'm surprised clippy was this smart.

File: src/web/mod.rs [hunk 7]
```diff
@@ -475,29 +475,29 @@ pub(crate) struct MetaData {
...
-        None
+        let rows = conn.query(
+        "SELECT crates.name,
+                       releases.version,
+                       releases.description,
+                       releases.target_name,
+                       releases.rustdoc_status,
+                       releases.default_target
+                FROM releases
+                INNER JOIN crates ON crates.id = releases.crate_id
```
Comment (jyn514) [inline src/web/mod.rs:486]: We probably should make this JOIN a view at some point ... but that's separate from your PR.

File: src/web/releases.rs [hunk 8]
```diff
@@ -398,14 +398,16 @@ pub fn author_handler(req: &mut Request) -> IronResult<Response> {
...
     let author = ctry!(router.find("author")
-        .ok_or(IronError::new(Nope::CrateNotFound, status::NotFound)));
-
-    let (author_name, packages) = if author.starts_with("@") {
-        let mut author = author.clone().split("@");
-        get_releases_by_owner(&conn,
-                              page_number,
-                              RELEASES_IN_RELEASES,
-                              cexpect!(author.nth(1)))
+        .ok_or_else(|| IronError::new(Nope::CrateNotFound, status::NotFound)));
```
Comment (jyn514) [inline src/web/releases.rs:401]: Is this change necessary? This isn't doing anything expensive and it adds mental overhead.

File: src/docbuilder/rustwide_builder.rs (hunk 2 above)
Comment (Kixiron) [reply src/docbuilder/rustwide_builder.rs:111]: It's actually a performance thing, `&str` has a fast, specialized `to_string` implementation while `&&str` does not

File: src/docbuilder/rustwide_builder.rs (hunk 3 above)
Comment (Kixiron) [reply src/docbuilder/rustwide_builder.rs:218]: Same thing with the first one, it's more performant to dereference

File: src/utils/release_activity_updater.rs (hunk 6 above)
Comment (Kixiron) [reply src/utils/release_activity_updater.rs:11]: Well my friend, that wasn't clippy

File: src/web/releases.rs (hunk 8 above)
Comment (Kixiron) [reply src/web/releases.rs:401]: It's actually a performance thing, `ok_or` always allocates, no matter the status of what it's called on, while `ok_or_else` is lazy, which means it's only allocated for when it's actually needed

File: src/docbuilder/mod.rs (hunk 1 above)
Comment (Kixiron) [reply src/docbuilder/mod.rs:67]: I'm not sure what the error type would be, or can I just `?`?

Comment (jyn514) [reply src/docbuilder/mod.rs:67]: Yes, `?` should work. All the error types are `failure::Error`, which implements `From` for most errors.

File: src/docbuilder/rustwide_builder.rs (hunk 2 above)
Comment (jyn514) [reply src/docbuilder/rustwide_builder.rs:111]: Oh interesting. Can you leave that as a comment and maybe a link to the documentation?

File: src/web/releases.rs (hunk 8 above)
Comment (jyn514) [reply src/web/releases.rs:401]: Right, but there's no allocation here. `IronError` is only on the stack, as is `Nope` and `NotFound`.

Comment (Kixiron) [reply src/web/releases.rs:401]: Alright, I'll revert that

File: src/docbuilder/rustwide_builder.rs (hunk 2 above)
Comment (Kixiron) [reply src/docbuilder/rustwide_builder.rs:111]: Best I've got is the [clippy lint](https://rust-lang.github.io/rust-clippy/master/index.html#inefficient_to_string), but it explains it fairly well. `&str` has it's own `ToString` impl, while `&&str` goes through `Display`, which is much slower

**Comment by jyn514** [comment]
Let's come back to this after the clippy and rustfmt PRs, so I'll be able to see what actually changed instead of looking at formatting.

File: src/docbuilder/rustwide_builder.rs [hunk 9]
```diff
@@ -122,8 +122,9 @@ impl RustwideBuilder {
 
         let mut targets_to_install = TARGETS
             .iter()
-            .map(|&t| t.to_string())
+            .map(|t| (*t).to_string()) // &str has a specialized ToString impl, while &&str goes through Display
             .collect::<HashSet<_>>();
```
Comment (jyn514) [inline src/docbuilder/rustwide_builder.rs:126]: I liked this better before. The comment is fine but I prefer `|&t|` to the parentheses.

File: src/docbuilder/rustwide_builder.rs [hunk 10]
```diff
@@ -225,7 +226,7 @@ impl RustwideBuilder {
                     let file_name = if versioned {
                         format!("{}-{}.{}", segments[1], rustc_version, segments[0])
                     } else {
-                        file.to_string()
+                        (*file).to_string() // &str has a specialized ToString impl, while &&str goes through Display
```
Comment (jyn514) [inline src/docbuilder/rustwide_builder.rs:229]: I don't think this dereference is necessary, `file` is already a `&str` because of the implicit dereference in the pattern on line 223.

File: src/utils/daemon.rs [hunk 11]
```diff
@@ -18,17 +18,15 @@ use time;
...
+    const CRATE_VARIABLES: [&str; 3] = [
         "CRATESFYI_PREFIX",
         "CRATESFYI_GITHUB_USERNAME",
         "CRATESFYI_GITHUB_ACCESSTOKEN",
-    ]
-    .iter()
-    {
-        if env::var(v).is_err() {
-            panic!("Environment variable {} not found", v);
-        }
```
Comment (jyn514) [inline src/utils/daemon.rs:31]: I prefer the way it was before, I think it was easier to read. Also I don't see the point of having a `const` since it's only used once, and that directly next to where the array is defined.

File: src/utils/release_activity_updater.rs [hunk 12]
```diff
@@ -25,22 +25,25 @@ pub fn update_release_activity() -> Result<()> {
...
+
         let release_count: i64 = rows.get(0).get(0);
         let failure_count: i64 = failures_count_rows.get(0).get(0);
         let now = now();
         let date = now - Duration::days(day);
+
+        // unwrap is fine here, as our date format is always valid
         dates.push(format!("{}", date.strftime("%d %b").unwrap()));
-        // unwrap is fine here,             ~~~~~~~~~~~~^  our date format is always valid
+
```
Comment (jyn514) [inline src/utils/release_activity_updater.rs:46]:
```suggestion
```

**Comment by jyn514** [comment]
Please update the PR description to reflect what this does, it's changed a lot since you opened it.

File: src/utils/daemon.rs (hunk 11 above)
Comment (Kixiron) [reply src/utils/daemon.rs:31]: I think it's much cleaner this way because of how it removes the rather large array away from the for statement

Comment (Kixiron) [reply src/utils/daemon.rs:31]: The const also just gives some more guarantees that it'll be flattened in the final code

Comment (jyn514) [reply src/utils/daemon.rs:31]: > I think it's much cleaner this way because of how it removes the rather large array away from the for statement

Ok, can you change it to just being a `const` array instead of `unwrap_or_else`? The first time I read the unwrap_or_else I was worried it panicked eagerly instead of on error :sweat_smile: 

> The const also just gives some more guarantees that it'll be flattened in the final code

Can you expand on this? My understanding is that for any `const X = blah;`, `X` behaves exactly as if you'd typed `blah` in the source code (modulo bugs in rustc).

Comment (Kixiron) [reply src/utils/daemon.rs:31]: > Ok, can you change it to just being a `const` array instead of `unwrap_or_else`? The first time I read the unwrap_or_else I was worried it panicked eagerly instead of on error 😅
Sure thing, it's actually better as the if now that I look at it

> Can you expand on this? My understanding is that for any` const X = blah;`, `X` behaves exactly as if you'd typed` blah` in the source code (modulo bugs in rustc).
Yah, that's exactly what I mean, the `const` makes sure that the variable is "the same" which could possibly not happen with a `let`. Adding onto the `const` train, if ever more variables are added then we don't end up with the longest for statement in the world, we just extend the const some. It gives more clarity to the code imo, but that's a personal thing
